src/components/CalendarContainer.ts

```diff
@@ -1,14 +1,215 @@
+            this.fetchEventsByXPath(mxObject.getGuid());
+        } else if (dataSource === "microflow" && dataSourceMicroflow) {
+            this.fetchEventsByMicroflow(mxObject);
+        }
+    }
+
+    private resetSubscriptions = (mxObject?: mendix.lib.MxObject) => {
+        this.subscriptionHandles.forEach(window.mx.data.unsubscribe);
+        this.subscriptionHandles = [];
+
+        if (mxObject) {
+            this.subscriptionHandles.push(window.mx.data.subscribe({
+                callback: () => this.fetchData(mxObject),
+                entity: this.props.eventEntity
+            }));
+            this.subscriptionHandles.push(window.mx.data.subscribe({
+                callback: () => this.fetchData(mxObject),
+                guid: mxObject.getGuid()
+            }));
+            [
+                this.props.titleAttribute,
+                this.props.startAttribute,
+                this.props.endAttribute
+            ].forEach(attr => this.subscriptionHandles.push(window.mx.data.subscribe({
+                attr,
+                callback: () => this.fetchData(mxObject), guid: mxObject.getGuid()
+            })));
+        }
+    }
+
+    private fetchEventsByXPath(contextGuid: string) {
+        const { entityConstraint, eventEntity } = this.props;
+        const constraint = entityConstraint ? entityConstraint.replace(/\[%CurrentObject%\]/g, contextGuid) : "";
+        const XPath = `//${eventEntity}${constraint}`;
+
+        window.mx.data.get({
+            xpath: XPath,
+            callback: mxObjects => this.setEventsFromMxObjects(mxObjects)
+        });
+    }
+
+    private setEventsFromMxObjects = (mxObjects: mendix.lib.MxObject[]) => {
+        const { titleAttribute, startAttribute, endAttribute } = this.props;
+        const events = mxObjects.map(mxObject => {
+            const title = mxObject.get(titleAttribute) as string;
+            const start = new Date(mxObject.get(startAttribute) as number);
+            const end = new Date(mxObject.get(endAttribute) as number);
+            const guid = mxObject.getGuid();
+
+            return {
+                title,
+                start,
+                end,
+                guid
+            };
+        });
+        this.setState({ events });
+    }
+
+    private fetchEventsByMicroflow(mxObject?: mendix.lib.MxObject) {
+        const { dataSourceMicroflow } = this.props;
+        if (mxObject && dataSourceMicroflow) {
+            window.mx.ui.action(dataSourceMicroflow, {
+                callback: (mxObjects: mendix.lib.MxObject[]) => this.setEventsFromMxObjects(mxObjects),
+                error: error => window.mx.ui.error(`Error while executing microflow: ${dataSourceMicroflow}: ${error.message}`),
+                params: {
+                    applyto: "selection",
+                    guids: mxObject ? [ mxObject.getGuid() ] : []
+                }
+            });
+        }
+    }
+
+    private excecuteEventAction = () => {
```

Comment:
i suggest u combine function `excecuteEventAction` and `excecuteSlotAction` cause they'all do the same thing.
HINT: if the difference between the 2 are `**arguments**` , then create 1 with needed `**params**`